### ISSUE TITLE
core/variant.h strict-aliasing warning with GCC 5

### ISSUE THREAD

**Issue by pgruenbacher** [godotengine/godot#23015]
**Godot version:**
fe0db6c4797b8094c64e4ad017e29be866288628 (master devel)

**OS**
Ubuntu 16.0

**Issue**
annoying warning on 

```
./core/variant.h: In member function 'Variant::ObjData& Variant::_get_obj()':
./core/variant.h:441:52: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *reinterpret_cast<ObjData *>(&_data._mem[0]);
                                                    ^
./core/variant.h: In member function 'const Variant::ObjData& Variant::_get_obj() const':
./core/variant.h:446:58: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *reinterpret_cast<const ObjData *>(&_data._mem[0]);
```

**Steps to reproduce:**
scons platform=x11 target=release-debug

**Comment by pgruenbacher** [comment]
`g++ -o editor/plugins/script_text_editor.x11.opt.tools.64.o -c -g1 -O2 -ffast-math -DDEBUG_ENABLED -pipe -Wall -Wno-sign-compare -Werror=return-type -DZSTD_STATIC_LINKING_ONLY -DFREETYPE_ENABLED -DSVG_ENABLED -DTOUCH_ENABLED -DALSA_ENABLED -DALSAMIDI_ENABLED -DPULSEAUDIO_ENABLED -DJOYDEV_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DOPENGL_ENABLED -DGLES_ENABLED -DGLAD_ENABLED -DGLES_OVER_GL -D_REENTRANT -DPTRCALL_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DHAVE_MNTENT -Ieditor -I. -I/usr/include/freetype2 -I/usr/include/libpng12 -Iplatform/x11 -Ithirdparty/zstd -Ithirdparty/glad -Ithirdparty/nanosvg editor/plugins/script_text_editor.cpp
In file included from ./core/method_ptrcall.h:36:0,
                 from ./core/method_bind.h:35,
                 from ./core/class_db.h:34,
                 from ./scene/main/node.h:34,
                 from ./scene/2d/canvas_item.h:34,
                 from ./scene/gui/control.h:36,
                 from ./scene/gui/container.h:34,
                 from ./scene/gui/box_container.h:34,
                 from ./scene/gui/color_picker.h:34,
                 from editor/plugins/script_text_editor.h:34,
                 from editor/plugins/script_text_editor.cpp:31:
./core/variant.h: In member function ‘Variant::ObjData& Variant::_get_obj()’:
./core/variant.h:441:52: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *reinterpret_cast<ObjData *>(&_data._mem[0]);
                                                    ^
./core/variant.h: In member function ‘const Variant::ObjData& Variant::_get_obj() const’:
./core/variant.h:446:58: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *reinterpret_cast<const ObjData *>(&_data._mem[0]);
`

**Comment by pgruenbacher** [comment]
```
g++ -o editor/plugins/script_text_editor.x11.opt.tools.64.o -c -g1 -O2 -ffast-math -DDEBUG_ENABLED -pipe -Wall -Wno-sign-compare -Werror=return-type -DZSTD_STATIC_LINKING_ONLY -DFREETYPE_ENABLED -DSVG_ENABLED -DTOUCH_ENABLED -DALSA_ENABLED -DALSAMIDI_ENABLED -DPULSEAUDIO_ENABLED -DJOYDEV_ENABLED -DX11_ENABLED -DUNIX_ENABLED -DOPENGL_ENABLED -DGLES_ENABLED -DGLAD_ENABLED -DGLES_OVER_GL -D_REENTRANT -DPTRCALL_ENABLED -DTOOLS_ENABLED -DGDSCRIPT_ENABLED -DMINIZIP_ENABLED -DHAVE_MNTENT -Ieditor -I. -I/usr/include/freetype2 -I/usr/include/libpng12 -Iplatform/x11 -Ithirdparty/zstd -Ithirdparty/glad -Ithirdparty/nanosvg editor/plugins/script_text_editor.cpp
In file included from ./core/method_ptrcall.h:36:0,
                 from ./core/method_bind.h:35,
                 from ./core/class_db.h:34,
                 from ./scene/main/node.h:34,
                 from ./scene/2d/canvas_item.h:34,
                 from ./scene/gui/control.h:36,
                 from ./scene/gui/container.h:34,
                 from ./scene/gui/box_container.h:34,
                 from ./scene/gui/color_picker.h:34,
                 from editor/plugins/script_text_editor.h:34,
                 from editor/plugins/script_text_editor.cpp:31:
./core/variant.h: In member function ‘Variant::ObjData& Variant::_get_obj()’:
./core/variant.h:441:52: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *reinterpret_cast<ObjData *>(&_data._mem[0]);
                                                    ^
./core/variant.h: In member function ‘const Variant::ObjData& Variant::_get_obj() const’:
./core/variant.h:446:58: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  return *reinterpret_cast<const ObjData *>(&_data._mem[0]);
```

**Comment by pgruenbacher** [comment]
```
g++ --version
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by pgruenbacher** [comment]
methods at fault. 
```Variant::ObjData &Variant::_get_obj() {

	return *reinterpret_cast<ObjData *>(&_data._mem[0]);
}

const Variant::ObjData &Variant::_get_obj() const {

	return *reinterpret_cast<const ObjData *>(&_data._mem[0]);
}
```

**Comment by pgruenbacher** [comment]
but my real complain is that since this function is inlined, this warning gets thrown for every implementation file compiled.

**Comment by pgruenbacher** [comment]
@Zylann you been compiling master lately? Have you been getting these warnings?

**Comment by Zylann** [comment]
I use MSVC, I get 361 warnings but not this one.

**Comment by pgruenbacher** [comment]
i don't think issue is that useful much anymore, I'll re-open later after 3.1 release if still an issue.

**Comment by akien-mga** [comment]
Warnings won't fix themselves if we don't keep issues open about them ;)

**Comment by akien-mga** [comment]
For the reference, those warnings are not raised with GCC 8 (at least on `scons p=x11 tools=yes target=debug`). But it would still be good to see if we can fix them for earlier GCC versions too, or if they're false positives.

**Comment by akien-mga** [comment]
The warnings are also raised on Travis CI with GCC 4.8, when building `p=server tools=yes target=release_debug`: https://travis-ci.org/akien-mga/godot/jobs/468953697

Interestingly, they're not raised in the `p=x11 tools=yes target=debug` and `p=x11 tools=no target=release` builds, so might actually be linked to `target=release_debug`.

**Comment by akien-mga** [comment]
Any idea @hpvb?

**Comment by marxin** [comment]
> For the reference, those warnings are not raised with GCC 8 (at least on `scons p=x11 tools=yes target=debug`). But it would still be good to see if we can fix them for earlier GCC versions too, or if they're false positives.

I wouldn't care about these warnings from ancient versions of GCC. What about closing this issues once #26027 is merged?

**Comment by akien-mga** [comment]
@marxin The problem is that our CI builds on Travis CI run on GCC 4.8 (Ubuntu 14.04), and those warnings make builds noticeably slower.

It looks like Travis *finally* provides an [Ubuntu 16.04 image](https://blog.travis-ci.com/2018-11-08-xenial-release), so I can give it a try.

**Comment by marxin** [comment]
> @marxin The problem is that our CI builds on Travis CI run on GCC 4.8 (Ubuntu 14.04), and those warnings make builds noticeably slower.

I see. I would enable that only on a recent GCC version.

> 
> It looks like Travis _finally_ provides an [Ubuntu 16.04 image](https://blog.travis-ci.com/2018-11-08-xenial-release), so I can give it a try.

Hm, which GCC is it providing? If I see correctly, Travis guys don't bother updating to latest Ubuntu:
https://github.com/hpc/charliecloud/issues/308

And there's no openSUSE release availabe in Travis. That's bad :-1:

**Comment by marxin** [comment]
But ```gcc-8```  should be available as seen here:
https://github.com/samtools/samtools/pull/952

**Comment by marxin** [comment]
@akien-mga Can we close this now?

**Comment by akien-mga** [comment]
Yeah, thanks.
